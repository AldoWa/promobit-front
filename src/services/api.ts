import axios from 'axios';

const {
  VITE_API_MOVIE_AUTHORIZATION: AUTHORIZATION,
} = import.meta.env;


const api = axios.create({
  baseURL: 'https://api.themoviedb.org/3/',
  headers: {
    accept: 'application/json',
    Authorization: `Bearer ${AUTHORIZATION}`,
  },
});

export default api;
