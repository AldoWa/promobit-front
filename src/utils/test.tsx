import { RenderOptions, render } from '@testing-library/react'

import { ChakraProvider } from '@chakra-ui/react'

interface AllTheProvidersProps {
  children: React.ReactNode
}

const AllTheProviders = ({children}: AllTheProvidersProps) => {
  return (
    <ChakraProvider>
      {children}
    </ChakraProvider>
  )
}

const customRender = (ui:React.ReactElement, options?: RenderOptions) =>
  render(ui, {wrapper: AllTheProviders, ...options})

// eslint-disable-next-line react-refresh/only-export-components
export * from '@testing-library/react'

// override render method
export {customRender as render}