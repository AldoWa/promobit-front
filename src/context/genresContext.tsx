import { createContext, useCallback, useEffect, useState } from 'react';

import api from '../services/api';

interface GenresProviderProps {
  children: React.ReactNode;
}

interface Genre {
  id: number;
  name: string;
}

export interface GenresResponse {
  genres: Genre[];
}

export interface GenresContextData {
  genres: Genre[];
  handleGenre: (id: number, type: 'add' | 'remove') => void;
  genresIDsSelected: number[];
  resetGenresSelected: () => void;
}

export const GenresContext = createContext<GenresContextData>(
  {} as GenresContextData
);
export function GenresProvider({ children }: GenresProviderProps) {
  const [genres, setGenres] = useState<Genre[]>([]);
  const [genresIDsSelected, setGenresSelected] = useState<number[]>([]);

  useEffect(() => {
    const getGenres = async () => {
      const { data } = await api.get<GenresResponse>('genre/movie/list?language=pt')

      setGenres(data.genres)
    }

    getGenres()
  }, [])

  const addGenreSelected = useCallback((id: number) => {
    setGenresSelected([...genresIDsSelected, id])
  }, [genresIDsSelected])

  const removeGenreSelected = useCallback((id: number) => {
    const newGenres = genresIDsSelected.filter(genre => genre !== id)
    setGenresSelected(newGenres)
  }, [genresIDsSelected])

  const resetGenresSelected = useCallback(() => setGenresSelected([]), [])

  const handleGenre = useCallback((id: number, type: 'add' | 'remove') => {
    if (type === 'add') {
      addGenreSelected(id)
      return;
    }
    removeGenreSelected(id)
  }, [addGenreSelected, removeGenreSelected])


  return (
    <GenresContext.Provider value={{ genres, genresIDsSelected, handleGenre, resetGenresSelected }}>
      {children}
    </GenresContext.Provider>
  );
}
