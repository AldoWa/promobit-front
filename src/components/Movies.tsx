import React, { useContext, useEffect, useState } from 'react';

import { Container, Grid, Text } from '@chakra-ui/react';

import MovieCard from './MovieCard';

import api from '../services/api';
import Pagination from './Pagination';
import { GenresContext } from '../context/genresContext';

interface Movie {
  title: string;
  poster_path: string;
  release_date: string;
  genre_ids: number[];
}

interface MoviesResponse {
  page: number;
  results: Movie[],
  total_pages: number;
}

const Movies: React.FC = () => {
  const [movies, setMovies] = useState<Movie[]>([])
  const [maxTotalPages,] = useState<number>(500)
  const [actualPage, setActualPage] = useState<number>(1)

  const { genresIDsSelected, resetGenresSelected } = useContext(GenresContext)

  useEffect(() => {
    const getMovies = async () => {
      const { data } = await api.get<MoviesResponse>(`movie/popular?language=pt-BR&page=1`)

      setMovies(data.results)
      setActualPage(data.page)
    }

    getMovies()
  }, [])

  const handlePage = async (page: number) => {
    resetGenresSelected()
    try {
      const { data } = await api.get<MoviesResponse>(`movie/popular?language=pt-BR&page=${page}`)

      setMovies(data.results)
      setActualPage(data.page)

   
    } catch (error) {
      console.log(error)
    }
  }

  const moviesFilter = (movie: Movie) => {
    if (genresIDsSelected.length === 0) return true
    return movie.genre_ids.some(genre => genresIDsSelected.includes(genre))
  }

  const moviesFiltered = movies.filter(movie => moviesFilter(movie))

  return (
    <Container maxW='1250px' mt={30}>
      {moviesFiltered.length === 0 ?
        <Text fontSize='xl' color='black'>
          Nenhum filme encontrado
        </Text> :
        <Grid templateColumns='repeat(6, 1fr)' gap='2rem' width='100%'>
          {moviesFiltered.map(movie => (
            <MovieCard
              key={movie.title}
              src={`https://image.tmdb.org/t/p/original${movie.poster_path}`}
              title={movie.title}
              year={movie.release_date}
            />
          ))}
        </Grid>
      }
      <Pagination
        handlePage={handlePage}
        maxTotalPages={maxTotalPages}
        actualPage={actualPage}
      />
    </Container>
  );
};

export default Movies;
