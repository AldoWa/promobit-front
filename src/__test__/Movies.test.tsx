import Movies from "../components/Movies";
import { GenresContextData } from "../context/genresContext";
import { fireEvent, render, screen, waitFor } from "../utils/test";
import api from "../services/api";

jest.mock('../services/api.ts')

const apiMocked = api as jest.Mocked<typeof api>

describe('Movies', () => {
  const providerProps: GenresContextData = {
    genres: [
      {
        id: 28,
        name: 'Action'
      },
      {
        id: 29,
        name: 'Drama'
      }
    ],
    resetGenresSelected: () => {},
    genresIDsSelected: [],
    handleGenre: () => {}
  }

  const moviesResponse = {
    results: [
      {
        "adult": false,
        "backdrop_path": "/r9oTasGQofvkQY5vlUXglneF64Z.jpg",
        "genre_ids": [
          28,
          35
        ],
        "id": 1029575,
        "original_language": "en",
        "original_title": "The Family Plan",
        "overview": "Dan Morgan is many things: a devoted husband, a loving father, a celebrated car salesman. He's also a former assassin. And when his past catches up to his present, he's forced to take his unsuspecting family on a road trip unlike any other.",
        "popularity": 3086.075,
        "poster_path": "/jLLtx3nTRSLGPAKl4RoIv1FbEBr.jpg",
        "release_date": "2023-12-14",
        "title": "The Family Plan",
        "video": false,
        "vote_average": 7.398,
        "vote_count": 600
      },
      {
        "adult": false,
        "backdrop_path": "/sRLC052ieEzkQs9dEtPMfFxYkej.jpg",
        "genre_ids": [
          878
        ],
        "id": 848326,
        "original_language": "en",
        "original_title": "Rebel Moon - Part One: A Child of Fire",
        "overview": "When a peaceful colony on the edge of the galaxy finds itself threatened by the armies of the tyrannical Regent Balisarius, they dispatch Kora, a young woman with a mysterious past, to seek out warriors from neighboring planets to help them take a stand.",
        "popularity": 2101.05,
        "poster_path": "/ui4DrH1cKk2vkHshcUcGt2lKxCm.jpg",
        "release_date": "2023-12-15",
        "title": "Rebel Moon - Part One: A Child of Fire",
        "video": false,
        "vote_average": 6.5,
        "vote_count": 1071
      },
    ],
    page: 1,
  }

  beforeEach(() => {
    apiMocked.get.mockResolvedValue({
      data: moviesResponse
    })
  })

  it('Should render the component and show one movie', async () => {
    render(<Movies />, { providerProps })

    const moviesGrid = await screen.findByTestId('movies')
    const movies = moviesGrid.childNodes
    expect(moviesGrid).toBeInTheDocument()
    expect(movies.length).toBe(2)
  })

  it('Should render only the movies with the genre selected', async () => {
    render(<Movies />, { providerProps: { ...providerProps, genresIDsSelected: [878] } })

    const moviesGrid = await screen.findByTestId('movies')
    const movies = moviesGrid.childNodes
    expect(moviesGrid).toBeInTheDocument()
    expect(movies.length).toBe(1)
  })

  it('Should paginate the movies', async () => {
    const resetGenresSelected = jest.fn()
    render(<Movies />, { providerProps: { ...providerProps, resetGenresSelected } })
    const paginationLastPage = await screen.findByTestId('pagination-last-page')
    fireEvent.click(paginationLastPage)
    await waitFor(() => expect(resetGenresSelected).toHaveBeenCalledTimes(1))
  })

  it("Should match snapshot", async () => {
    const { container } = render(<Movies />, { providerProps })
    await waitFor(() => expect(container).toMatchSnapshot())
  });
});
