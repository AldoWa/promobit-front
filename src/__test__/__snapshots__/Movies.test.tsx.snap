// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies Should match snapshot 1`] = `
<div>
  <div
    class="chakra-container css-7sj89i"
  >
    <p
      class="chakra-text css-1mva2b3"
    >
      Nenhum filme encontrado
    </p>
    <div
      class="chakra-stack css-7e2ce"
      data-testid="pagination-stack"
    >
      <div
        class="chakra-stack css-1igwmid"
        data-testid="pagination-pages"
      >
        <button
          aria-labelledby="Ir para a página"
          class="chakra-button css-1s4k305"
          type="button"
        >
          1
        </button>
        <button
          aria-labelledby="Ir para a página"
          class="chakra-button css-1ie58vi"
          type="button"
        >
          2
        </button>
        <button
          aria-labelledby="Ir para a página"
          class="chakra-button css-1ie58vi"
          type="button"
        >
          3
        </button>
        <button
          aria-labelledby="Ir para a página"
          class="chakra-button css-1ie58vi"
          type="button"
        >
          4
        </button>
        <button
          aria-labelledby="Ir para a página"
          class="chakra-button css-1ie58vi"
          type="button"
        >
          5
        </button>
      </div>
      <button
        class="chakra-button css-ez23ye"
        type="button"
      >
        &gt;
      </button>
      <button
        class="chakra-button css-ez23ye"
        data-testid="pagination-last-page"
        type="button"
      >
        Última
      </button>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
